diff -ur -x .build -x .inst -x .sinst -x back-java -x CYGWIN-PATCHES openldap-2.2.26-orig/build/top.mk openldap-2.2.26/build/top.mk
--- openldap-2.2.26-orig/build/top.mk	2006-07-19 11:49:51.109375000 -0500
+++ openldap-2.2.26/build/top.mk	2006-07-18 21:50:25.703125000 -0500
@@ -185,13 +185,14 @@
 MODULES_CPPFLAGS = @SLAPD_MODULES_CPPFLAGS@
 MODULES_LDFLAGS = @SLAPD_MODULES_LDFLAGS@
 MODULES_LIBS = @MODULES_LIBS@
+SLAPD_JAVA_LDFLAGS = @SLAPD_JAVA_LDFLAGS@
 SLAPD_PERL_LDFLAGS = @SLAPD_PERL_LDFLAGS@
 
 SLAPD_SQL_LDFLAGS = @SLAPD_SQL_LDFLAGS@
 SLAPD_SQL_INCLUDES = @SLAPD_SQL_INCLUDES@
 SLAPD_SQL_LIBS = @SLAPD_SQL_LIBS@
 
-SLAPD_LIBS = @SLAPD_LIBS@ @SLAPD_PERL_LDFLAGS@ @SLAPD_SQL_LDFLAGS@ @SLAPD_SQL_LIBS@ @SLAPD_SLP_LIBS@
+SLAPD_LIBS = @SLAPD_LIBS@ @SLAPD_JAVA_LDFLAGS@ @SLAPD_PERL_LDFLAGS@ @SLAPD_SQL_LDFLAGS@ @SLAPD_SQL_LIBS@ @SLAPD_SLP_LIBS@
 SLURPD_LIBS = @SLURPD_LIBS@
 
 # Our Defaults
diff -ur -x .build -x .inst -x .sinst -x back-java -x CYGWIN-PATCHES openldap-2.2.26-orig/configure openldap-2.2.26/configure
--- openldap-2.2.26-orig/configure	2006-07-19 11:49:51.171875000 -0500
+++ openldap-2.2.26/configure	2006-07-18 21:58:20.015625000 -0500
@@ -107,6 +107,8 @@
 ac_help="$ac_help
     --enable-passwd	  enable passwd backend no|yes|mod [no]"
 ac_help="$ac_help
+    --enable-java	  enable java backend no|yes|mod [no]"
+ac_help="$ac_help
     --enable-perl	  enable perl backend no|yes|mod [no]"
 ac_help="$ac_help
     --enable-shell	  enable shell backend no|yes|mod [no]"
@@ -1858,7 +1860,7 @@
 # end --enable-wrappers
 
 Backends="bdb dnssrv hdb ldap ldbm meta monitor null \
-	passwd perl shell sql"
+	passwd java perl shell sql"
 
 # Check whether --with-xxslapoverlays or --without-xxslapoverlays was given.
 if test "${with_xxslapoverlays+set}" = set; then
@@ -2089,6 +2091,26 @@
   	ol_enable_passwd="no"
 fi
 # end --enable-passwd
+# OpenLDAP --enable-java
+        # Check whether --enable-java or --disable-java was given.
+if test "${enable_java+set}" = set; then
+  enableval="$enable_java"
+
+        ol_arg=invalid
+        for ol_val in no yes mod ; do
+                if test "$enableval" = "$ol_val" ; then
+                        ol_arg="$ol_val"
+                fi
+        done
+        if test "$ol_arg" = "invalid" ; then
+                { echo "configure: error: bad value $enableval for --enable-java" 1>&2; exit 1; }
+        fi
+        ol_enable_java="$ol_arg"
+
+else
+        ol_enable_java="no"
+fi
+# end --enable-java
 # OpenLDAP --enable-perl
 	# Check whether --enable-perl or --disable-perl was given.
 if test "${enable_perl+set}" = set; then
@@ -2365,6 +2387,7 @@
 		$ol_enable_monitor = no -a \
 		$ol_enable_null = no -a \
 		$ol_enable_passwd = no -a \
+		$ol_enable_java = no -a \
 		$ol_enable_perl = no -a \
 		$ol_enable_shell = no -a \
 		$ol_enable_sql = no ; then
@@ -2473,6 +2496,7 @@
 BUILD_MONITOR=no
 BUILD_NULL=no
 BUILD_PASSWD=no
+BUILD_JAVA=no
 BUILD_PERL=no
 BUILD_SHELL=no
 BUILD_SQL=no
@@ -2488,6 +2512,10 @@
 SLAPD_STATIC_BACKENDS=
 SLAPD_DYNAMIC_BACKENDS=
 
+SLAPD_JAVA_LDFLAGS=
+MOD_JAVA_LDFLAGS=
+JAVA_CPPFLAGS=
+
 SLAPD_PERL_LDFLAGS=
 MOD_PERL_LDFLAGS=
 PERL_CPPFLAGS=
@@ -7493,6 +7521,89 @@
 fi
 
 
+ol_link_java=no
+if test $ol_enable_java != no ; then
+
+        echo =====================================================================================
+        echo ol_enable_java = $ol_enable_java
+
+        # Extract the first word of "java", so it can be a program name with args.
+set dummy java; ac_word=$2
+echo $ac_n "checking for $ac_word""... $ac_c" 1>&6
+echo "configure:7502: checking for $ac_word" >&5
+if eval "test \"\${ac_cv_path_JAVABIN+set}\" = set"; then
+  echo $ac_n "(cached) $ac_c" 1>&6
+else
+  case "$JAVABIN" in
+  /*)
+  ac_cv_path_JAVABIN="$JAVABIN" # Let the user override the test with a path.
+  ;;
+  ?:/*)
+  ac_cv_path_JAVABIN="$JAVABIN" # Let the user override the test with a dos path.
+  ;;
+  *)
+  IFS="${IFS=   }"; ac_save_ifs="$IFS"; IFS=":"
+  ac_dummy="$PATH"
+  for ac_dir in $ac_dummy; do
+    test -z "$ac_dir" && ac_dir=.
+    if test -f $ac_dir/$ac_word; then
+      ac_cv_path_JAVABIN="$ac_dir/$ac_word"
+      break
+    fi
+  done
+  IFS="$ac_save_ifs"
+  test -z "$ac_cv_path_JAVABIN" && ac_cv_path_JAVABIN="/usr/local/java/bin/java"
+  ;;
+esac
+fi
+JAVABIN="$ac_cv_path_JAVABIN"
+if test -n "$JAVABIN"; then
+  echo "$ac_t""$JAVABIN" 1>&6
+else
+  echo "$ac_t""no" 1>&6
+fi
+
+
+        if test "no$JAVABIN" = "no" ; then
+                if test $ol_enable_java = yes ; then
+                        { echo "configure: error: could not locate java" 1>&2; exit 1; }
+                fi
+
+        else
+                case $host_os in
+                cygwin* | mingw* | pw32*)
+                        JAVA_HOME=`/usr/bin/cygpath --unix $JAVA_HOME`
+                        JAVA_CPPFLAGS="-D__int64='long long' -I$JAVA_HOME/include -I$JAVA_HOME/include/win32"
+                        JAVA_LDFLAGS="-L`pwd`/servers/slapd/back-java -L$JAVA_HOME/jre/bin -L$JAVA_HOME/jre/bin/server -ljava -ljvm -lverify"
+                        ;;
+                darwin* | rhapsody*)
+                        JAVA_CPPFLAGS="-I/System/Library/Frameworks/JavaVM.framework/Headers"
+                        JAVA_LDFLAGS="-framework JavaVM"
+
+                        ;;
+                *)
+                        JAVA_CPPFLAGS="-I$JAVA_HOME/include -I$JAVA_HOME/include/linux"
+                        JAVA_LDFLAGS="-L$JAVA_HOME/jre/lib/i386 -L$JAVA_HOME/jre/lib/i386/server  -L$JAVA_HOME/jre/lib/i386 -ljava -ljvm -lverify"
+                        ;;
+                esac
+
+                if test x"$ol_enable_java" = "xyes" ; then
+                        SLAPD_JAVA_LDFLAGS="$JAVA_LDFLAGS"
+                else
+                        MOD_JAVA_LDFLAGS="$JAVA_LDFLAGS"
+                fi
+                                ol_link_java=yes
+        fi
+
+        echo JAVA_HOME          = $JAVA_HOME
+        echo JAVA_CPPFLAGS      = $JAVA_CPPFLAGS
+        echo SLAPD_JAVA_LDFLAGS = $SLAPD_JAVA_LDFLAGS
+        echo MOD_JAVA_LDFLAGS   = $MOD_JAVA_LDFLAGS
+
+        echo =====================================================================================
+
+fi
+
 ol_link_perl=no
 if test $ol_enable_perl != no ; then
 	# Extract the first word of "perl", so it can be a program name with args.
@@ -24131,6 +24237,22 @@
 
 fi
 
+if test "$ol_link_java" != no ; then
+        BUILD_SLAPD=yes
+        BUILD_JAVA=$ol_enable_java
+        if test "$ol_enable_java" = mod ; then
+                SLAPD_DYNAMIC_BACKENDS="$SLAPD_DYNAMIC_BACKENDS back-java"
+                MFLAG=SLAPD_MOD_DYNAMIC
+        else
+                SLAPD_STATIC_BACKENDS="$SLAPD_STATIC_BACKENDS back-java"
+                MFLAG=SLAPD_MOD_STATIC
+        fi
+        cat >> confdefs.h <<EOF
+#define SLAPD_JAVA $MFLAG
+EOF
+
+fi
+
 if test "$ol_link_perl" != no ; then
 	BUILD_SLAPD=yes
 	BUILD_PERL=$ol_enable_perl
@@ -24399,6 +24521,7 @@
 servers/slapd/back-monitor/Makefile:build/top.mk:servers/slapd/back-monitor/Makefile.in:build/mod.mk \
 servers/slapd/back-null/Makefile:build/top.mk:servers/slapd/back-null/Makefile.in:build/mod.mk \
 servers/slapd/back-passwd/Makefile:build/top.mk:servers/slapd/back-passwd/Makefile.in:build/mod.mk \
+servers/slapd/back-java/Makefile:build/top.mk:servers/slapd/back-java/Makefile.in:build/mod.mk \
 servers/slapd/back-perl/Makefile:build/top.mk:servers/slapd/back-perl/Makefile.in:build/mod.mk \
 servers/slapd/back-shell/Makefile:build/top.mk:servers/slapd/back-shell/Makefile.in:build/mod.mk \
 servers/slapd/back-sql/Makefile:build/top.mk:servers/slapd/back-sql/Makefile.in:build/mod.mk \
@@ -24490,6 +24613,7 @@
 s%@CPP@%$CPP%g
 s%@LIBTOOL@%$LIBTOOL%g
 s%@LN_H@%$LN_H%g
+s%@JAVABIN@%$JAVABIN%g
 s%@PERLBIN@%$PERLBIN%g
 s%@OL_MKDEP@%$OL_MKDEP%g
 s%@OL_MKDEP_FLAGS@%$OL_MKDEP_FLAGS%g
@@ -24511,6 +24635,7 @@
 s%@BUILD_MONITOR@%$BUILD_MONITOR%g
 s%@BUILD_NULL@%$BUILD_NULL%g
 s%@BUILD_PASSWD@%$BUILD_PASSWD%g
+s%@BUILD_JAVA@%$BUILD_JAVA%g
 s%@BUILD_PERL@%$BUILD_PERL%g
 s%@BUILD_SHELL@%$BUILD_SHELL%g
 s%@BUILD_SQL@%$BUILD_SQL%g
@@ -24530,6 +24655,9 @@
 s%@SLAPD_STATIC_BACKENDS@%$SLAPD_STATIC_BACKENDS%g
 s%@SLAPD_DYNAMIC_BACKENDS@%$SLAPD_DYNAMIC_BACKENDS%g
 s%@SLAPD_DYNAMIC_OVERLAYS@%$SLAPD_DYNAMIC_OVERLAYS%g
+s%@JAVA_CPPFLAGS@%$JAVA_CPPFLAGS%g
+s%@SLAPD_JAVA_LDFLAGS@%$SLAPD_JAVA_LDFLAGS%g
+s%@MOD_JAVA_LDFLAGS@%$MOD_JAVA_LDFLAGS%g
 s%@PERL_CPPFLAGS@%$PERL_CPPFLAGS%g
 s%@SLAPD_PERL_LDFLAGS@%$SLAPD_PERL_LDFLAGS%g
 s%@MOD_PERL_LDFLAGS@%$MOD_PERL_LDFLAGS%g
@@ -24616,6 +24744,7 @@
 servers/slapd/back-monitor/Makefile:build/top.mk:servers/slapd/back-monitor/Makefile.in:build/mod.mk \
 servers/slapd/back-null/Makefile:build/top.mk:servers/slapd/back-null/Makefile.in:build/mod.mk \
 servers/slapd/back-passwd/Makefile:build/top.mk:servers/slapd/back-passwd/Makefile.in:build/mod.mk \
+servers/slapd/back-java/Makefile:build/top.mk:servers/slapd/back-java/Makefile.in:build/mod.mk \
 servers/slapd/back-perl/Makefile:build/top.mk:servers/slapd/back-perl/Makefile.in:build/mod.mk \
 servers/slapd/back-shell/Makefile:build/top.mk:servers/slapd/back-shell/Makefile.in:build/mod.mk \
 servers/slapd/back-sql/Makefile:build/top.mk:servers/slapd/back-sql/Makefile.in:build/mod.mk \
diff -ur -x .build -x .inst -x .sinst -x back-java -x CYGWIN-PATCHES openldap-2.2.26-orig/configure.in openldap-2.2.26/configure.in
--- openldap-2.2.26-orig/configure.in	2005-03-14 11:06:26.000000000 -0600
+++ openldap-2.2.26/configure.in	2006-07-18 21:49:38.921875000 -0500
@@ -198,7 +198,7 @@
 dnl ----------------------------------------------------------------
 dnl SLAPD Backend Options
 Backends="bdb dnssrv hdb ldap ldbm meta monitor null \
-	passwd perl shell sql"
+	passwd java perl shell sql"
 
 AC_ARG_WITH(xxslapoverlays,[
 SLAPD Backend Options:])
@@ -216,6 +216,7 @@
 OL_ARG_ENABLE(monitor,[    --enable-monitor	  enable monitor backend no|yes|mod], yes, [no yes mod])dnl
 OL_ARG_ENABLE(null,[    --enable-null	  enable null backend no|yes|mod], no, [no yes mod])dnl
 OL_ARG_ENABLE(passwd,[    --enable-passwd	  enable passwd backend no|yes|mod], no, [no yes mod])dnl
+OL_ARG_ENABLE(java,[    --enable-java	  enable java backend no|yes|mod], no, [no yes mod])dnl
 OL_ARG_ENABLE(perl,[    --enable-perl	  enable perl backend no|yes|mod], no, [no yes mod])dnl
 OL_ARG_ENABLE(shell,[    --enable-shell	  enable shell backend no|yes|mod], no, [no yes mod])dnl
 OL_ARG_ENABLE(sql,[    --enable-sql	  enable sql backend no|yes|mod], no, [no yes mod])dnl
@@ -332,6 +333,7 @@
 		$ol_enable_monitor = no -a \
 		$ol_enable_null = no -a \
 		$ol_enable_passwd = no -a \
+		$ol_enable_java = no -a \
 		$ol_enable_perl = no -a \
 		$ol_enable_shell = no -a \
 		$ol_enable_sql = no ; then
@@ -444,6 +446,7 @@
 BUILD_MONITOR=no
 BUILD_NULL=no
 BUILD_PASSWD=no
+BUILD_JAVA=no
 BUILD_PERL=no
 BUILD_SHELL=no
 BUILD_SQL=no
@@ -459,6 +462,10 @@
 SLAPD_STATIC_BACKENDS=
 SLAPD_DYNAMIC_BACKENDS=
 
+SLAPD_JAVA_LDFLAGS=
+MOD_JAVA_LDFLAGS=
+JAVA_CPPFLAGS=
+
 SLAPD_PERL_LDFLAGS=
 MOD_PERL_LDFLAGS=
 PERL_CPPFLAGS=
@@ -556,6 +563,56 @@
 AC_PROG_LN_S
 
 dnl ----------------------------------------------------------------
+dnl Java
+ol_link_java=no
+if test $ol_enable_java != no ; then
+
+        echo =====================================================================================
+        echo ol_enable_java = $ol_enable_java
+
+        AC_PATH_PROG(JAVABIN, java, /usr/local/java/bin/java)
+
+        if test "no$JAVABIN" = "no" ; then
+                if test $ol_enable_java = yes ; then
+                        AC_MSG_ERROR([could not locate java])
+                fi
+
+        else
+                case $host_os in
+                cygwin* | mingw* | pw32*)
+                        JAVA_HOME=`/usr/bin/cygpath --unix $JAVA_HOME`
+                        JAVA_CPPFLAGS="-D__int64='long long' -I$JAVA_HOME/include -I$JAVA_HOME/include/win32"
+                        JAVA_LDFLAGS="-L`pwd`/servers/slapd/back-java -L$JAVA_HOME/jre/bin -L$JAVA_HOME/jre/bin/server -ljava -ljvm -lverify"
+                        ;;
+                *)
+                        JAVA_CPPFLAGS="-I$JAVA_HOME/include -I$JAVA_HOME/include/linux"
+                        JAVA_LDFLAGS="-L$JAVA_HOME/jre/lib/i386 -L$JAVA_HOME/jre/lib/i386/server -ljava -ljvm -lverify
+
+                        ;;
+                esac
+
+                if test x"$ol_enable_java" = "xyes" ; then
+                        SLAPD_JAVA_LDFLAGS="$JAVA_LDFLAGS"
+                else
+                        MOD_JAVA_LDFLAGS="$JAVA_LDFLAGS"
+                fi
+                dnl should check java version
+                ol_link_java=yes
+        fi
+
+        echo JAVA_HOME          = $JAVA_HOME
+        echo JAVA_CPPFLAGS      = $JAVA_CPPFLAGS
+        echo JAVA_LDFLAGS       = $JAVA_LDFLAGS
+        echo SLAPD_JAVA_LDFLAGS = $SLAPD_JAVA_LDFLAGS
+        echo MOD_JAVA_LDFLAGS   = $MOD_JAVA_LDFLAGS
+
+        echo =====================================================================================
+
+fi
+
+AC_PROG_CPP
+
+dnl ----------------------------------------------------------------
 dnl Perl
 ol_link_perl=no
 if test $ol_enable_perl != no ; then
@@ -2525,6 +2582,19 @@
 	AC_DEFINE_UNQUOTED(SLAPD_PASSWD,$MFLAG,[define to support PASSWD backend])
 fi
 
+if test "$ol_link_java" != no ; then
+        BUILD_SLAPD=yes
+        BUILD_JAVA=$ol_enable_java
+        if test "$ol_enable_java" = mod ; then
+                SLAPD_DYNAMIC_BACKENDS="$SLAPD_DYNAMIC_BACKENDS back-java"
+                MFLAG=SLAPD_MOD_DYNAMIC
+        else
+                SLAPD_STATIC_BACKENDS="$SLAPD_STATIC_BACKENDS back-java"
+                MFLAG=SLAPD_MOD_STATIC
+        fi
+        AC_DEFINE_UNQUOTED(SLAPD_JAVA,$MFLAG,[define to support JAVA backend])
+fi
+
 if test "$ol_link_perl" != no ; then
 	BUILD_SLAPD=yes
 	BUILD_PERL=$ol_enable_perl
@@ -2644,6 +2714,7 @@
   AC_SUBST(BUILD_MONITOR)
   AC_SUBST(BUILD_NULL)
   AC_SUBST(BUILD_PASSWD)
+  AC_SUBST(BUILD_JAVA)
   AC_SUBST(BUILD_PERL)
   AC_SUBST(BUILD_SHELL)
   AC_SUBST(BUILD_SQL)
@@ -2668,6 +2739,10 @@
 AC_SUBST(SLAPD_DYNAMIC_BACKENDS)
 AC_SUBST(SLAPD_DYNAMIC_OVERLAYS)
 
+AC_SUBST(JAVA_CPPFLAGS)
+AC_SUBST(SLAPD_JAVA_LDFLAGS)
+AC_SUBST(MOD_JAVA_LDFLAGS)
+
 AC_SUBST(PERL_CPPFLAGS)
 AC_SUBST(SLAPD_PERL_LDFLAGS)
 AC_SUBST(MOD_PERL_LDFLAGS)
@@ -2726,6 +2801,7 @@
 servers/slapd/back-monitor/Makefile:build/top.mk:servers/slapd/back-monitor/Makefile.in:build/mod.mk \
 servers/slapd/back-null/Makefile:build/top.mk:servers/slapd/back-null/Makefile.in:build/mod.mk \
 servers/slapd/back-passwd/Makefile:build/top.mk:servers/slapd/back-passwd/Makefile.in:build/mod.mk \
+servers/slapd/back-java/Makefile:build/top.mk:servers/slapd/back-java/Makefile.in:build/mod.mk \
 servers/slapd/back-perl/Makefile:build/top.mk:servers/slapd/back-perl/Makefile.in:build/mod.mk \
 servers/slapd/back-shell/Makefile:build/top.mk:servers/slapd/back-shell/Makefile.in:build/mod.mk \
 servers/slapd/back-sql/Makefile:build/top.mk:servers/slapd/back-sql/Makefile.in:build/mod.mk \
diff -ur -x .build -x .inst -x .sinst -x back-java -x CYGWIN-PATCHES openldap-2.2.26-orig/include/portable.h.in openldap-2.2.26/include/portable.h.in
--- openldap-2.2.26-orig/include/portable.h.in	2005-03-14 11:06:26.000000000 -0600
+++ openldap-2.2.26/include/portable.h.in	2006-07-18 21:50:55.171875000 -0500
@@ -974,6 +974,9 @@
 /* define to support PASSWD backend */
 #undef SLAPD_PASSWD
 
+/* define to support JAVA backend */
+#undef SLAPD_JAVA
+
 /* define to support PERL backend */
 #undef SLAPD_PERL
 
diff -ur -x .build -x .inst -x .sinst -x back-java -x CYGWIN-PATCHES openldap-2.2.26-orig/servers/slapd/backend.c openldap-2.2.26/servers/slapd/backend.c
--- openldap-2.2.26-orig/servers/slapd/backend.c	2005-01-20 11:01:06.000000000 -0600
+++ openldap-2.2.26/servers/slapd/backend.c	2006-07-18 21:51:45.890625000 -0500
@@ -83,6 +83,9 @@
 #if SLAPD_PASSWD == SLAPD_MOD_STATIC
 #include "back-passwd/external.h"
 #endif
+#if SLAPD_JAVA == SLAPD_MOD_STATIC
+#include "back-java/external.h"
+#endif
 #if SLAPD_PERL == SLAPD_MOD_STATIC
 #include "back-perl/external.h"
 #endif
@@ -130,6 +133,9 @@
 #if SLAPD_PASSWD == SLAPD_MOD_STATIC
 	{"passwd",	passwd_back_initialize},
 #endif
+#if SLAPD_JAVA == SLAPD_MOD_STATIC
+	{"java",	java_back_initialize},
+#endif
 #if SLAPD_PERL == SLAPD_MOD_STATIC
 	{"perl",	perl_back_initialize},
 #endif
